---
title: "VCFAnnotator-Tempus: A tool to annotate the variants supplied in a VCF file"
author: |
    | Ashish Jain
    | Bioinformatics and Computational Biology
    | Iowa State University, Ames, Iowa
    | 
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
output: 
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    #number_sections: true  ## if you want number sections at each table header
    #theme: united  # many options for theme, this one is my favorite.
fig_width: 8 
fig_height: 5 
vignette: >
  %\VignetteIndexEntry{TissueEnrich}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  fig.width=8, fig.height=5
)
```

# VCFAnnotator-Tempus
This is an R package for the automatic annotation generation of the variants supplied in the VCF file into a text file.

**Requirement**

You need R version 3.5.1 or above to run this application. Other dependencies are:

* `tidyverse (>= 1.2.1)`
* `ensurer (>= 1.1.0)`
* `httr (>= 1.3.1)`
* `jsonlite (>=1.6.0)`

**How to install the R package**

* Install Dependencies

```{r eval=FALSE}
install.packages(c("tidyverse","ensurer","httr","jsonlite"))
```

**From Github**

```{r eval=FALSE}
install.packages("devtools")
library("devtools")
install_github("ashishjain1988/VCFAnnotator-Tempus")
```


**From tar file**

```{r eval=FALSE}
install.packages("<Absolute Path>/VCFAnnotatorTempus_1.0.0.tar.gz", repos = NULL)
```

**Running the VCF Annotator**

VCF Annotator is an R package used for the automatic annotation generation of the variants supplied in the VCF file. This package parses information of variants supplied in a VCF file. This package also add the annotations from the Broad Institue's ExAC project by making a POST call to its [batch API](http://exac.hms.harvard.edu/rest/bulk/variant). The output in the text file have the annotations including 

* Type of variation (most deleterious)
* Depth of sequence coverage
* Number of reads supporting variants
* Percentage of reads supporting the variant versus those supporting reference reads
* Allele frequency from ExAC project
* Consequence of the variant from ExAC project

In our code, we rank the type of variations according to the magnitude of its deleterious possibilty (Insertion,Deletion > Complex > Multi Nucleotide Polymorphism > Single Nucleotide Polymorphism). In order to annotate the variants supplied in the VCF file `annotateVariant` function of this package is used. This function requires the path of the VCF file and then automatically annotates the variants and supply them in a data.frame object which can be written into a text file. A simple example is shown below by using an R terrminal.

```{r eval=FALSE}
library(VCFAnnotatorTempus)

VCFfilePath <- system.file('extdata', 'Challenge_data.vcf', package = 'VCFAnnotatorTempus')
t <- annotateVariant(file = VCFfilePath)
write.table(t,"Challenge_data_annotated_variants.txt",sep="\t",quote = F,row.names = F)
```
In addition to that, there is another function `getVariantInfoFromExACAPI` which make a batch call to the Broad Institue's ExAC project API to retreive the annotation information of variants. More information about this function can be found by typing `?getVariantInfoFromExACAPI` in R terminal. This function is called by the `annotateVariant` function internally to add annotations from ExAC project.
